File: java/ToolGood/Words/StringSearchEx.java
package ToolGood.Words;

import java.util.ArrayList;
import java.util.List;

import ToolGood.Words.internals.BaseSearchEx;

public class StringSearchEx extends BaseSearchEx{


    public List<String> FindAll(String text)
    {
        List<String> root = new ArrayList<String>();
        int p = 0;

        for (int i = 0; i < text.length(); i++) {
            int t =  _dict[text.charAt(i)];
            if (t == 0) {
                p = 0;
                continue;
            }
            int next = _next[p] + t;
            boolean find = _key[next] == t;
            if (find == false && p != 0) {
                p = 0;
                next = _next[0] + t;
                find = _key[next] == t;
            }
            if (find) {
                int index = _check[next];
                if (index > 0) {
                    for (int item : _guides[index]) {
                        root.add(_keywords[item]);
                    }
                }
                p = next;
            }
        }
        return root;
    }

    public String FindFirst(String text)
    {
        Integer p = 0;
        for (int i = 0; i < text.length(); i++) {
            int t =  _dict[text.charAt(i)];
            if (t == 0) {
                p = 0;
                continue;
            }
            int next = _next[p] + t;
            if (_key[next] == t) {
                int index = _check[next];
                if (index > 0) {
                    return _keywords[_guides[index][0]];
                }
                p = next;
            } else {
                p = 0;
                next = _next[p] + t;
                if (_key[next] == t) {
                    int index = _check[next];
                    if (index > 0) {
                        return _keywords[_guides[index][0]];
                    }
                    p = next;
                }
            }
        }
        return null;
    }


    public boolean ContainsAny(String text)
    {
        Integer p = 0;
        for (int i = 0; i < text.length(); i++) {
            int t =  _dict[text.charAt(i)];
            if (t == 0) {
                p = 0;
                continue;
            }
            int next = _next[p] + t;
            if (_key[next] == t) {
                if (_check[next] > 0) { return true; }
                p = next;
            } else {
                p = 0;
                next = _next[p] + t;
                if (_key[next] == t) {
                    if (_check[next] > 0) { return true; }
                    p = next;
                }
            }
        }
        return false;
    }

    public String Replace(String text){
        return Replace(text,'*');
    }

    public String Replace(String text, Character replaceChar)
    {
        StringBuilder result = new StringBuilder(text);

        Integer p = 0;

        for (int i = 0; i < text.length(); i++) {
            int t =  _dict[text.charAt(i)];
            if (t == 0) {
                p = 0;
                continue;
            }
            int next = _next[p] + t;
            boolean find = _key[next] == t;
            if (find == false && p != 0) {
                p = 0;
                next = _next[p] + t;
                find = _key[next] == t;
            }
            if (find) {
                int index = _check[next];
                if (index > 0) {
                    int maxLength = _keywords[_guides[index][0]].length();
                    int start = i + 1 - maxLength;
                    for (int j = start; j <= i; j++) {
                        result.setCharAt(j, replaceChar);
                     }
                 }
                 p = next;
            }
        }
        return result.toString();
    }
}

File: golang/ToolGood/Words/StringSearchEx.go
package Words

import . "./internals"

 
type StringSearchEx struct {
	BaseSearchEx
}

func NewStringSearchEx() *StringSearchEx{
	return &StringSearchEx{
	}
}

func (this *StringSearchEx)FindAll(text string) []string  {
	root:= make([]string, 0)
	p:=0
	for _,c := range text {
		t:=this.I_dict[c]
		if	t==0{
			p = 0;
			continue;
		}
		next := this.I_next[p] + t;
		find := this.I_key[next] == t;
		if  find == false && p != 0  {
			p = 0;
			next = this.I_next[0] + t;
			find = this.I_key[next] == t;
		}
		if  find  {
			index := this.I_check[next];
			if (index > 0) {
				for _,item:=range this.I_guides[index]{
					root=append(root,this.I_keywords[item])
				}
			}
			p = next;
		}
	}
	return root;
}
 

func (this *StringSearchEx)FindFirst(text string) string  {
	p:=0
	for _,c := range text {
		t:=this.I_dict[c]
		if	t==0{
			p = 0;
			continue;
		}
		next := this.I_next[p] + t;
		find := this.I_key[next] == t;
		if  find == false && p != 0  {
			p = 0;
			next = this.I_next[0] + t;
			find = this.I_key[next] == t;
		}
		if  find  {
			index := this.I_check[next];
			if (index > 0) {
				return this.I_keywords[this.I_guides[index][0]];
			}
			p = next;
		}
	}
	return "";
}
 

func (this *StringSearchEx)ContainsAny(text string) bool  {
	p:=0
	for _,c := range text {
		t:=this.I_dict[c]
		if	t==0{
			p = 0;
			continue;
		}
		next := this.I_next[p] + t;

		if this.I_key[next] == t {
			if  this.I_check[next] > 0  { 
				return true;
			}
			p = next;
		} else {
			p = 0;
			next = this.I_next[p] + t;
			if (this.I_key[next] == t) {
				if (this.I_check[next] > 0) {
					 return true;
				}
				p = next;
			}
		}
	}
	return false;
}
 

func (this *StringSearchEx)Replace(text string, replaceChar rune) string  {
	result:= []rune (text)
	p:=0
	var i int
	for _,c := range text {
		t:=this.I_dict[c]
		if	t==0{
			p = 0;
			i++
			continue;
		}
		next := this.I_next[p] + t;
		find := this.I_key[next] == t;
		if  find == false && p != 0  {
			p = 0;
			next = this.I_next[0] + t;
			find = this.I_key[next] == t;
		}
		if  find  {
			index := this.I_check[next];
			if (index > 0) {
				r:=this.I_keywords[this.I_guides[index][0]]
				maxLength:=len([]rune(r))
				start:= i + 1 - maxLength
				for j := start; j <= i; j++{
					result[j]=replaceChar
				} 
			}
			p = next;
		}
		i++
	}
	return string (result) 
}
